clang-tidy: enable more checks, run on GitHub Actions

Octol1ttle opened the pull request:
This PR enables most of checks provided by clang-tidy, with checks that do not fit us excluded. The table containing naming conventions was using wrong values, that was fixed and now tidy correctly warns and fixes naming convention issues
clang-tidy is now also ran on changed files within pull requests. Warnings are displayed on the pull request itself similar to CodeQL

Octol1ttle committed:
clang-tidy: enable more checks, run on GitHub Actions

Signed-off-by: Octol1ttle <l1ttleofficial@outlook.com>

Trial97 reviewed (approved):
Well wouldn't it be better to do this in nix?

Octol1ttle commented:
> Well wouldn't it be better to do this in nix?

i have zero nix knowledge

getchoo commented on .github/workflows/build.yml:
```diff
@@ -159,7 +159,29 @@ jobs:
       - name: Configure project
         run: |
-          cmake --preset "$CMAKE_PRESET"
+          cmake --preset "$CMAKE_PRESET" -DCMAKE_EXPORT_COMPILE_COMMANDS=ON
```
```suggestion
          cmake --preset "$CMAKE_PRESET"
```
https://github.com/PrismLauncher/PrismLauncher/blob/ff98c350a1c09f84faf7a2e08af7c50175f23b62/CMakeLists.txt#L110-L114

getchoo commented on .github/workflows/build.yml:
```diff
@@ -159,7 +159,29 @@ jobs:
+          cmake --preset "$CMAKE_PRESET" -DCMAKE_EXPORT_COMPILE_COMMANDS=ON
+
+      - name: Run clang-tidy
+        if: ${{ github.event_name == 'pull_request' }}
```
I don't think we need to *only* run this on PRs

getchoo commented on .github/workflows/build.yml:
```diff
@@ -159,7 +159,29 @@ jobs:
+          fi
+
+      - name: Show diff of clang-tidy autofixes
+        if: always()
```
This causes the above to run even if clang-tidy never did

getchoo commented on .github/workflows/build.yml:
```diff
@@ -159,7 +159,29 @@ jobs:
+            FILES=$(gh -R ${{ github.repository }} pr diff ${{ github.event.pull_request.number }} --name-only | grep -E '\.(cpp|h)$' || true)
+          
+            if [ -n "$FILES" ]; then
+              echo "$FILES" | xargs clang-tidy -p=build/ --fix-errors --quiet --warnings-as-errors='*'
```
I don't think we need to do this in a Nix flake check or anything, but we *should* be using `clang-tidy` from our development shell here to keep it in sync with the LLVM version we use for other tooling (we've had some weird discrepancies between versions across CI and dev machines)

getchoo reviewed (commented):
This is something I've been meaning to do for a while, so thanks for the push

I think this would be best best left to another workflow, though. I don't think we really need to be running this across the entire matrix, and it's not exactly a part of our build in the first place

We'd likely also be much better off using GitHub's [SARIF support](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning) and a tool like [clang-tidy-sarif](https://crates.io/crates/clang-tidy-sarif), as this would give us review-like lints similar to CodeQL and automatically(?) handle lints caught across multiple workflows. This would hopefully make the lint errors more apparent and help us avoid the brittleness of *only* linting files that changed

Octol1ttle committed:
change(actions/clang-tidy): also run if config file was changed

Signed-off-by: Octol1ttle <l1ttleofficial@outlook.com>

Octol1ttle committed:
clang-tidy: remove options that don't exist

Signed-off-by: Octol1ttle <l1ttleofficial@outlook.com>

Octol1ttle committed:
include checks related to High Integrity C++ Coding Standard

Signed-off-by: Octol1ttle <l1ttleofficial@outlook.com>

Octol1ttle committed:
clang-tidy: disable requiring all 5 special member functions

Signed-off-by: Octol1ttle <l1ttleofficial@outlook.com>

Octol1ttle committed:
actions(clang-tidy): don't autofix errors

Signed-off-by: Octol1ttle <l1ttleofficial@outlook.com>

Octol1ttle committed:
clang-tidy: TIL about check aliases

Signed-off-by: Octol1ttle <l1ttleofficial@outlook.com>

Octol1ttle committed:
clang-tidy: allow magic numbers

Signed-off-by: Octol1ttle <l1ttleofficial@outlook.com>

abhicommands reviewed (approved):
This finally means clang-tidy latest version wont fail Great!

abhicommands commented:
@Octol1ttle I found an issue, the libnbtplusplus on libraries is also being checked for clang-format, clang-tidy is being ignored but not clang-format is there a way to resolve this? I am not too familiar, but I believe doing this fixes it. 


+DisableFormat: true
on clangformat for libnbtplusplus will send a pr for that on ur fork!

DioEgizio reviewed (approved):
i wonder if we could add [clazy](https://invent.kde.org/sdk/clazy#clang-tidy) checks too

Octol1ttle committed:
clang-tidy: disable easily-swappable-parameters

Signed-off-by: Octol1ttle <l1ttleofficial@outlook.com>

Octol1ttle committed:
clang-tidy: re-enable performance-unnecessary-value-param

Pass-by-const-ref is used even in Qt code and is one of the enabled changes in #5095

Signed-off-by: Octol1ttle <l1ttleofficial@outlook.com>

Octol1ttle commented:
Superseded by #5051
